src/main/java/co/cask/publisher/PubSubOutputFormat.java

```diff
@@ -0,0 +1,218 @@
+  private static final String DELAY_THRESHOLD = "delay.threshold";
+  private static final String ERROR_THRESHOLD = "error.threshold";
+  private static final String RETRY_TIMEOUT_SECONDS = "retry.timeout";
+
+  public static void configure(Configuration configuration, GooglePublisher.Config config) {
+    configuration.set(SERVICE_PATH, config.serviceAccountFilePath);
+    String projectId = GCPUtil.getProjectId(config.project);
+    configuration.set(PROJECT, projectId);
+    configuration.set(TOPIC, config.topic);
+    configuration.set(COUNT_BATCH_SIZE, String.valueOf(config.getMessageCountBatchSize()));
+    configuration.set(REQUEST_BYTES_THRESHOLD, String.valueOf(config.getRequestBytesThreshold()));
+    configuration.set(DELAY_THRESHOLD, String.valueOf(config.getPublishDelayThresholdMillis()));
+    configuration.set(ERROR_THRESHOLD, String.valueOf(config.getErrorThreshold()));
+    configuration.set(RETRY_TIMEOUT_SECONDS, String.valueOf(config.getRetryTimeoutSeconds()));
+  }
+
+  @Override
+  public RecordWriter<NullWritable, Text> getRecordWriter(TaskAttemptContext taskAttemptContext) throws IOException {
+    Configuration config = taskAttemptContext.getConfiguration();
+    String serviceAccountFilePath = config.get(SERVICE_PATH);
+    String projectId = config.get(PROJECT);
+    String topic = config.get(TOPIC);
+    long countSize = Long.parseLong(config.get(COUNT_BATCH_SIZE));
+    long bytesThreshold = Long.parseLong(config.get(REQUEST_BYTES_THRESHOLD));
+    long delayThreshold = Long.parseLong(config.get(DELAY_THRESHOLD));
+    long errorThreshold = Long.parseLong(config.get(ERROR_THRESHOLD));
+    int retryTimeout = Integer.parseInt(config.get(RETRY_TIMEOUT_SECONDS));
+    Publisher publisher = Publisher.newBuilder(ProjectTopicName.of(projectId, topic))
+      .setCredentialsProvider(() -> GCPUtils.loadServiceAccountCredentials(serviceAccountFilePath))
+      .setBatchingSettings(getBatchingSettings(countSize, bytesThreshold, delayThreshold))
+      .setRetrySettings(getRetrySettings(retryTimeout))
+      .build();
+
+    return new PubSubRecordWriter(publisher, errorThreshold);
+  }
+
+  private RetrySettings getRetrySettings(int maxRetryTimeout) {
+    // only maxRetryTimeout is configurable by user, setting defaults for other parameters
+    return RetrySettings.newBuilder()
+      .setInitialRetryDelay(Duration.ofMillis(100))
+      .setRetryDelayMultiplier(2)
+      .setMaxRetryDelay(Duration.ofSeconds(2))
+      .setTotalTimeout(Duration.ofSeconds(maxRetryTimeout))
+      .setInitialRpcTimeout(Duration.ofSeconds(1))
+      .setMaxRpcTimeout(Duration.ofSeconds(5))
+      .build();
+  }
+
+  private BatchingSettings getBatchingSettings(long countSize, long bytesThreshold, long delayThreshold) {
+    BatchingSettings.Builder batchingSettingsBuilder = BatchingSettings.newBuilder();
+    batchingSettingsBuilder.setElementCountThreshold(countSize);
```

Comment:
this builder doesn't support chaining these set calls?

Reply:
it does, updated it.